api/src/main/java/net/wesjd/anvilgui/version/VersionMatcher.java

```diff
@@ -18,22 +16,40 @@ public class VersionMatcher {
      * @throws IllegalStateException If the version wrapper failed to be instantiated or is unable to be found
      */
     public VersionWrapper match() {
-        final String serverVersion = Bukkit.getServer()
-                .getClass()
-                .getPackage()
-                .getName()
-                .split("\\.")[3]
-                .substring(1);
-        try {
-            return (VersionWrapper) Class.forName(getClass().getPackage().getName() + ".Wrapper" + serverVersion)
-                    .getDeclaredConstructor()
-                    .newInstance();
-        } catch (ClassNotFoundException exception) {
-            throw new IllegalStateException(
-                    "AnvilGUI does not support server version \"" + serverVersion + "\"", exception);
-        } catch (ReflectiveOperationException exception) {
+        final String classPattern = "Wrapper%s_%s%s";
+        final String packageName = getClass().getPackage().getName();
+
+        final int major = this.versionProvider.major();
+        final int minor = this.versionProvider.minor();
+        final int patch = this.versionProvider.patch();
+
+        // Start searching for suitable VersionWrapper
+        // starting from the current patch version going downward
+        VersionWrapper suitableWrapper = null;
+        for (int i = patch; i >= 0; --i) {
```

Comment:
This would stop searching at the lowest patch version. Which works fine so far, because each new minor version thus far has required a new AnvilGUI module, due to Spigot's "R-version" packages.

But if Spigot stops using "R-version" packages in the future (which would not surprise me), we could end up in a situation where we have a Wrapper for, let's say, Minecraft 1.25, which also works for Minecraft 1.26. Then we would not be able to find the 1.25 Wrapper class, because the search stops at 1.26.

Honestly this is purely hypothetical at this point and doesn't really matter (we can fix it later if the problem arises), but I just wanted to highlight it for discussion.

The reason I bring this up is that I am wondering if it would be simpler (and also more future-proof) to just define an ordered list of Wrapper names (`1_8`, `1_8_3`, `1_8_8`, `...`, `1_20_5`) and then have the `VersionMatcher` try to load the highest supported version from the list. Just start from the back of the list and continue until a supported version is found (one that is less than or equal to the current version), then load that module. This is just an idea that came to my mind, let me know what you think.

Reply:
not quite sure what you are talking about. the lowest patch version would be zero, and if it's zero, I don't use the patch version, and instead I just use major_minor pattern. I really don't understand what can cause an issue here from what you are saying. because we have 1.10 and 1.11 versions as well. so what's the difference between 1.10–1.11 and 1.25–1.26? since we're going to separate 1.25 and 1.26 versions as WrapperV1_25 and WrapperV1_26 like how we did it in 1.10–1.11 versions.
the main reason why I don't like defining every single version in a map is because it'll require changing the map values on every single release. so I don't like putting new things that can go wrong while reviewing. making things as automatic as possible would be much more cleaner imo.

Reply:
> The problem with your current solution is that (unless I'm mistaken) it would start searching for `1_26_3`, then for `1_26_2`, `1_26_1` and finally for `1_26`. After that, it would give up the search and `throw new IllegalStateException("No compatible version wrapper found for server version v" + major + "." + minor + "." + patch);`. But that's not correct, because _there is_ a compatible version wrapper (`1_25_2`).

that's not a problem tho, we have to provide wrapper versions for each minecraft version. if there is no 1_26 wrapper which means we don't support 1.26 version yet. this is already how the library works right now. why we jump from 1_25 to 1_28? that doesn't make sense and never gonna happen either way.

Reply:
> this is already how the library works right now. why we jump from 1_25 to 1_28? that doesn't make sense and never gonna happen either way.

That *can* happen in the future, if Spigot stops using the "R-versions" and just uses Mojang mappings (same as Paper), which wouldn't surprise me. When that happens, it is totally possible for a 1.25.2 wrapper to be compatible with Minecraft 1.26, 1.27 etc. This can already be done for Paper, as mentioned by @mastercake10 in #294:

> I see one huge advantage in method 2), which is that the new wrapper class can be re-used for all upcoming paper versions, since it doesn't rely on imports with cb version numbers in it. Also mojang mappings doesn't change over releases.

Reply:
This is why we discussed just checking if the package name contains an R-Version, and if it does not, load the module through a lookup table in. We can't predict the future. As long as this project supports spigot, i don't really see the benefit of dropping the R-Version entirely. We haven't discussed it before, so I don't see why Paper's decision suddenly makes us doing this. We really just need to make it compatible with Paper again. This is also the consensus we reached in #294.

There is so much discussion going on here, we simply don't know what Spigot's next moves are, so why not just take the simple approach and see what the future brings?

Reply:
it doesn't make a difference if spigot is also remove revisions. we can always add support for new versions simply by adding more nms modules. there won't be any issue even if spigot decides to go down on the same road as paper.

Reply:
> That is not the same situation. There is a WrapperV1_10 and a WrapperV1_11, so your current solution will work for these versions because the search never has to continue down to a lower minor version. As illustrated above, the hypothetical case I described above is different in that there is a WrapperV1_25_2, which is also compatible with 1.26.x versions, but there are no WrapperV1_26_X version wrappers, so your current algorithm will not find a wrapper if the current version is 1.26.x even though there is a compatible one (WrapperV1_25_2).


if there is no 1_26_x version wrapper, then that means we don't support 1.26 yet, because we didn't create version wrapper for it?
i really don't understand what you guys are trying to say here. because if you think about it, if you add the version wrapper specifically for 1.26 then it will work, what's the problem here?

Reply:
> it doesn't make a difference if spigot removes revisions. we can always add support for new versions simply by adding more nms modules. there won't be any issue even if spigot decides to go down on the same road as paper.

Right, that's why we don't even need to drop the R at this point.

> i really don't understand what you guys are trying to say here. because if you think about it, if you add the version wrapper specifically for 1.26 then it will work, what's the problem here?

It assumes that we have wrappers for mojang-mapped servers (which is the case for paper 1.20.5+). The new wrapper class for 1.26 _could_ exactly be the same file for 1.25, so it might be as well just be reused instead of adding the same file again. Of course with Spigot mappings, the method names etc. change every release, which is why we need to add a new wrapper for every MC version.

Reply:
so the problem is not the mappings or R or anything, just DIY?

Reply:
we can create a base class for it so that would be really easy solution.

Reply:
> Right, that's why we don't even need to drop the R at this point.

i think we can do that but this is not the PR for it. maybe after refactoring the build system to gradle.

Reply:
> so the problem is not the mappings or R or anything, just DIY?

I don't think there is a problem for now, its just based on assumptions for the future. 

It comes down to personal preference, for me this whole process of finding the right Wrapper class for a given MC version is not really required, especially for < MC 1.20.5, where you can still look up the R-Version.

> i think we can do that but this is not the PR for it. maybe after refactoring the build system to gradle.

I didn't mean the module names, you already dropped the R in the class names. I have already a fork for gradle support as discussed [here](https://github.com/WesJD/AnvilGUI/issues/294#issuecomment-2082700651), but it doesn't drop the R (see the comment why).

Reply:
i think we should think about all of these whenever we need, let's just focus on merging this pr. sent request.

Reply:
> There is so much discussion going on here, we simply don't know what Spigot's next moves are, so why not just take the simple approach and see what the future brings?

> i think we should think about all of these whenever we need,

Once again, I brought this stuff with `VersionMatcher` up only for discussion, I agree that this is not something we have to fix right now.

> This is why we discussed just checking if the package name contains an R-Version, and if it does not, load the module through a lookup table in.

> It comes down to personal preference, for me this whole process of finding the right Wrapper class for a given MC version is not really required, especially for < MC 1.20.5, where you can still look up the R-Version.

This is the second point that I was trying to make, which is that the current `VersionMatcher` implementation (apart from possibly breaking in the future if Spigot moves in the direction of Paper and uses Mojang mappings without "R-versions") seems unnecessarily complex when we just simply just have a lookup table or list.

---

Once again, I don't oppose the current `VersionMatcher` implementation, it will definitely work. Ultimately the decision is not up to me either way, I am not a maintainer here, just throwing my ideas around.

Reply:
Good point @0dinD, but I think we should wait to fix this issue until the problem arises.